ckanext/opendata_theme/opengov_custom_header/controller.py

```diff
@@ -0,0 +1,184 @@
+        self.link = six.text_type(link)
+        self.position = position
+        self._html = None
+        self.html = six.text_type(html)
+        self.active = active
+
+    def __repr__(self):
+        return '{}:{}'.format(self.position, self.title)
+
+    def to_dict(self):
+        return {
+            'title': self.title,
+            'link': self.link,
+            'position': self.position,
+            'html': self._html,
+            'active': self.active,
+        }
+
+    @property
+    def html(self):
+        if not self._html:
+            self._html = literal('<li><a href="{}">{title}</a></li>'.format(
+                self.link, title=self.title))
+        return literal(self._html)
+
+    @html.setter
+    def html(self, value):
+        self._html = value
+
+
+class CustomHeaderController(admin.AdminController):
+    def remove_link(self):
+        if request.method == 'POST':
+            header_data = self.get_custom_header_metadata()
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            item = [link for link in header_data['links'] if link.title == data['to_remove']]
+            header_data['links'].remove(item[0])
+            error = self.save_header_metadata(header_data)
+            header_data['errors'] = error
+            self.redirect_to_custom_header_page(header_data)
+
+    def add_link(self):
+        if request.method == 'POST':
+            header_data = self.get_custom_header_metadata()
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            header_data['links'].append(
+                Header(
+                    title=data.get('new_title'),
+                    link=data.get('new_link'),
+                    position=len(header_data['links'])+1,
+                ))
+            error = self.save_header_metadata(header_data)
+            header_data['errors'] = error
+            return render('admin/custom_header.html',
+                          extra_vars=header_data)
+
+    def custom_header(self):
+        custom_header = self.get_custom_header_metadata()
+        if not custom_header.get('links'):
+            # this block is required for base initialization
+            # it happens only once when default custom header metadata is not set
+            # because it is set in build_pages_nav_main helper function which is called
+            # during the page rendering.
+            # reset_custom_header is able to render the page in the background for setting default metadata.
+            self.reset_custom_header()
+        if request.method == 'POST':
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            custom_header = {
+                'links': [],
+                'layout_type': data.get('layout_type', 'default')
+            }
+            data_len = len(data.get('link', []))
```

Comment:
There is an edge case if only one nav link is submitted from the front end. If there is only one url the form data will send `link` as a string instead of a list. Only when there are multiple links will `link` be a list.

Reply:
refactored to isinstance instead of len of field.